dev-docs/RFCs/v6.2/mapbox-custom-layer-rfc.md

```diff
@@ -0,0 +1,191 @@
+# RFC: Mapbox Custom Layer Integration
+
+* Authors: Xiaoji Chen, Ib Green
+* Date: Sep 25, 2018
+* Status: **Draft**
+
+References:
+
+* [Layer Group Operations RFC](v6.x/layer-and-group-operation-rfc.md)
+
+
+## Abstract
+
+This RFC proposes changes to the deck.gl API to allow deck.gl to use mapbox context and to trigger rendering of certain deck.gl layer renders in between mapbox style layers.
+
+
+## Problems
+
+* Initializing the DeckGL instance with an external gl context. (Work towards making this as declarative as possible).
+* Render deck.gl "layer-by-layer" - Work towards an official "static" rendering API for deck.gl
+
+
+## Proposed MapboxGL API Integration
+
+This case is fairly simple: a `MapboxLayer` class that can be used as a custom mapbox layer provides the plumbing and enables mapbox users to use deck.gl "layer packs".
+
+The adapter class transparently creates and manages as `Deck` instance under the hood.
+
+```js
+import Deck, {ScatterplotLayer} from '@deck.gl/core';
+import {MapboxLayer} from '@deck.gl/mapbox';
+
+const map = new mapboxgl.Map({
+  container: 'map',
+  style: 'mapbox://styles/mapbox/light-v9',
+  center: [-74.012, 40.705],
+  zoom: 15.5,
+  bearing: -20,
+  pitch: 45
+});
+
+map.on('load', () => {
+  map.addLayer(mapboxBuildingLayer);
+  map.addLayer(deckPoiLayer, getFirstTextLayerId(map.getStyle()));
+  map.addLayer(deckRouteLayer);
+});
+```
+
+
+## Proposed DeckGL API Integration
+
+This case is a bit harder since it both requires postponing deck.gl initialization.
+
+### Proposal 1: Callbacks
+
+```js
+import Deck, {ScatterplotLayer} from '@deck.gl/core';
+import {MapboxLayer} from '@deck.gl/mapbox';
+
+const map = new mapboxgl.Map({...});
+
+const deck = new Deck({
+  width: '100%',
+  height: '100%',
+  longitude: -122.4,
+  latitude: 37.8,
+  layers: [
+    new MapboxLayer({
+      id: 'my-scatterplot',
+      insertBefore: 'admin-1-labels',
+      type: ScatterplotLayer,
+      ...
+    })
+  ],
+  onAddLayer: (layer) => {
+    if (layer instanceof MapboxLayer) {
+      map.addLayer(layer, layer.props.insertBefore)
+    }
+  },
+  onRemoveLayer: (layer) => {
+    if (layer instanceof MapboxLayer) {
+      map.removeLayer(layer.id);
+    }
+  }
+});
+```
+
+Advantages:
+
+* Allows deck.gl and mapbox layers to be ordered (to some degree) within the deck.gl layer list
+* Callbacks are generic and could possibly be useful in other contexts
+
+Downsides:
+
+* Uses `MapboxLayer` inside deck.gl API, ideally that module should be reserved for mapbox API.
```

Comment:
This idea can be further explained with https://github.com/uber/deck.gl/pull/2306 - `MapboxLayer` inherits the base `Layer` class and `Deck` does not have any special knowledge about it.

Reply:
> MapboxLayer inherits the base Layer class and Deck does not have any special knowledge about it.

OK so this is a different `MapboxLayer` than the one used for the mapbox API integration? 

While deck.gl might be technically unaware, I do feel that mapbox related things gets more involved in the coding of the deck.gl API (mapbox layer, mapbox callbacks), whereas the alternative proposal keep things more separate and more loosely coupled. Both approaches have pros and cons, I suspect.

Reply:
```js
import React from 'react';
import DeckGL, {ScatterplotLayer} from 'deck.gl';
import {StaticMap} from 'react-map-gl';
import {MapboxLayer} from '@deck.gl/mapbox';

class App extends React.Component {

  state = {
    isMapLoaded: false
  }

  _renderLayers() {
    if (!isMapLoaded) {
      return [];
    }
    return [
      new MapboxLayer({
        id: 'my-scatterplot',
        insertBefore: getFirstTextLayerId(this._map.getStyle()),
        type: ScatterplotLayer,
        ...
      })
    ];
  }

  _onAddLayer = (layer) => {
    if (layer instanceof MapboxLayer) {
      this._map.addLayer(layer, layer.props.insertBefore);
    }
  }

  _onRemoveLayer = (layer) => {
    if (layer instanceof MapboxLayer) {
      this._map.removeLayer(layer.id);
    }
  }
  
  render() {
    return (
      <DeckGL
        initialViewState={...}
        controller={true}
        layers={this._renderLayers()}
        onAddLayer={this._onAddLayer}
        onRemoveLayer={this._onRemoveLayer}
      >
        <StaticMap
          ref={ref => {
            this._map = ref.getMap();
          }}
          mapboxApiAccessToken={...}
          onLoad={() => this.setState({isMapLoaded: true})} />
      </DeckGL>
    );
  }
}
```

Reply:
The same `MapboxLayer` implement both deck.gl `Layer` and Mapbox custom layer.

deck.gl has no mapbox-specific API in the above PR. The mapbox-related callbacks are defined by the application.